docs/zkapps/advanced-snarkyjs.mdx

```diff
@@ -291,3 +291,170 @@ We imagine that one of the most important use cases for actions is to enable con
+only storing one single hash on-chain!
+
+#### How does that work?
+
+Merkle trees are special binary trees in which every leaf (the nodes at the very bottom of the tree!) are cryptographic hashes of the underlying pieces of data and the internal nodes are labelled with the cryptographic hash of the concatenated labels (hashes) of its child nodes.
+
+By following this algorithm to the very top, we end up with one single node (the root node) that stores the root hash of the tree. The root hash is a reference to all pieces of data that were included in the tree's leaves. By doing so, we achieve that we can reference large amounts of data by only using one small hash!Another thing that Merkle trees provide is something called the witness (sometimes also called Merkle proof or Merkle path) - the witness is the path from one specific leaf node to the very top of the tree - to the root! Merkle witnesses are proofs of inclusion exists, they allow us to prove that one specific piece of data(an account in the ledger or the scores on a leaderboard) exists within the entire tree.
+
+#### How will that be useful for zkApps?
+
+We just learned that we can reference large amounts of off-chain data and prove inclusion of very specific parts of that data only with a small hash - the root - and something called a witness.
+
+When we want to use Merkle trees and reference off-chain data in our zkApps on Mina, we can simply store the root of the tree on-chain and voilà,
+we now have access to more data off-chain!
+
+Let's imagine a zkApp that manages a game with a leaderboard. The zkApp has a method to update the score of a player, if the player guesses a number correctly.
+Once a player reaches a threshold score, they can invoke another method to get a reward. Because we want many players to participate in our game,
+we are drastically limited by how much data we can store on-chain - and if we want to have at least 8 or more participants, we will run out of on-chain space very fast!
+
+A possible solution to that problem would be to utilize the power of Merkle trees and store the public keys of each player and their corresponding scores off-chain and only reference it in our smart contract.
+
+Lets look at our data structure first. Imagine wanting to map a player's id to score points, so this:
+
+```
+0: 5 points
+1: 3 points
+2: 0 points
+3: 8 points
+... : ...
+7: 2 points
+```
+
+#### Implementing the smart contract
+
+Now it's time to look at what our leaderboard zkApp might look like. We want to have on-chain state that points to the off-chain Merkle tree - we will call this variable the `root`.
+
+<Alert kind="note">
+
+Sometimes the variable `root` is also called commitment, because we commit to something!
+
+</Alert>
+
+Additionally, we want to store a variable `z` which is the hash of the value a player has to guess: `H(guess) = z`
```

Comment:
This can be brute force attacked b/c the user can fetch the zkApp account state and rainbow attack it. Ideally our example wouldn't have this deficiency. But if we can't find an easy such example, then let's add a follow up sentence or note box mentioning this deficiency and guiding to only this for toy use cases b/c it's not secure. But ideally we'd show something secure

Reply:
Ye, thats a very good point. I think its okay for this example to add a disclaimer 

```
Guessing a simple hash like in this example can easily be brute forced, especially if the preimage is something simple (like a 5 letter word or a small number with only a few digits).

Please make sure that your zkApps are always secure, especially when dealing with funds!

```